### ISSUE TITLE
Ego conditioned models

### ISSUE THREAD

**Issue by kensukenk** [NVlabs/traffic-behavior-simulation#8]
Hello, I was wondering if the current repository supports ego-action conditioned prediction for its implementation of TPP. I see in the DiscreteVAEConfig class that 'ego_conditioning' is set to False, 'EC_feat_dim' is set to 64, and 'eval_class' set to TPP. This leads me to believe that the DiscreteVAE model that TPP is using should be able to support ego conditioning if I set the 'ego_conditioning' flag to True. 

However, when I look at the RasterizedDiscreteCVAE, it doesn't appear to use alg_config.ego_conditioning anywhere. If I look further down to the RasterizedTreeVAE model, there is a conditional based on self.ego_conditioning. If I wanted to use an ego-conditioned variant of TPP like in the original paper, would the change be in a similar vein to what is in the RasterizedTreeVAE model?

**Comment by chenyx09** [comment]
Hi, we haven't tested ego-conditioned TPP before, it is definitely possible to replace the original TPP with an ego-conditioned version, but in our experience, models like TPP that are trained as prediction models and evaluated in closed-loop typically do not perform well.

**Comment by kensukenk** [comment]
Thank you! I also was trying to train the ego-conditioned agentformer model on the nuscenes dataset with discrete latent variables like in the tree-structured policy planning paper, but I keep running into Not Implemented errors in the agent_aware_attention. I was wondering if you ever ran into this issue? It was running fine for the first 30k iterations but now it randomly seems to run into this issue.

**Comment by chenyx09** [comment]
Hi, Kensukenk,
I tested agentformer training with ego-conditioning on, and it worked on my machine. What specific NotImplementedError were you seeing?
Yuxiao

**Comment by kensukenk** [comment]
Hi Yuxiao,

My error came here: https://github.com/NVlabs/traffic-behavior-simulation/blob/2f8edb4c7401d0bba0fe7277690803c9057eadcb/tbsim/models/agentformer_lib.py#L312C6-L313C38

I saw a similar issue came up here: https://github.com/pytorch/pytorch/issues/47979 so i tried this fix. I also tried using the query_no_nan/key_no_nan/value_no_nan version of the conditional that were commented out. In those cases, I would then get an error in the context encoder here: https://github.com/NVlabs/traffic-behavior-simulation/blob/2f8edb4c7401d0bba0fe7277690803c9057eadcb/tbsim/models/agentformer.py#L229

These errors happen very sporadically. Sometimes I can train the model for ~8 hours and it is fine, other times the error comes up after 45 minutes.

**Comment by chenyx09** [comment]
I see. I've never encountered this. My educated guess is some entries of the tensor become nan and it causes the check performed by torch.equal to fail. There are many possible reasons for the nan issue, a bad data point, a learning rate that is too high, some bad scaling, etc. My suggestion is to check when the nan happens and figure out why.

**Comment by kensukenk** [comment]
In case anyone else runs into this issue, on my end it seems that training fp16 caused some instability that led to NaN issues. I switched to single-precision and haven't run into these issues since.
